File: scripts/reconcile_task_loop_topics.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Reconcile Task Loop Topics
Auto-provisions topic thread sessions for enabled topics declared in topics.json.
"""

import sys
import json
import uuid
import argparse
from pathlib import Path
from datetime import datetime, timezone


def reconcile_topics(project_root: str, manifest_path: str, registry_path: str) -> dict:
    root = Path(project_root).resolve()
    manifest_file = Path(manifest_path) if Path(manifest_path).is_absolute() else root / manifest_path
    registry_file = Path(registry_path) if Path(registry_path).is_absolute() else root / registry_path

    if not manifest_file.is_file():
        return {"action": "INVALID", "reason": f"Manifest file not found: {manifest_file}"}
    if not registry_file.is_file():
        return {"action": "INVALID", "reason": f"Registry file not found: {registry_file}"}

    with open(manifest_file, "r", encoding="utf-8") as f:
        manifest = json.load(f)
    with open(registry_file, "r", encoding="utf-8") as f:
        registry = json.load(f)

    if "modules" not in registry:
        registry["modules"] = {}

    provisioned = []
    modified = False

    for topic in manifest.get("topics", []):
        if not topic or (topic.get("enabled") is False):
            continue

        mod_key = topic.get("module_key")
        if not mod_key:
            continue

        reg_mod = registry["modules"].get(mod_key)
        if not reg_mod or not reg_mod.get("thread_id"):
            new_thread_id = str(uuid.uuid4())
            registry["modules"][mod_key] = {
                "thread_id": new_thread_id,
                "title": topic.get("title", f"Topic {mod_key}"),
                "title_prefix": topic.get("title_prefix", f"{mod_key}-"),
                "memory_docs": topic.get("memory_docs", []),
                "auto_provisioned": True,
                "provisioned_at": datetime.now(timezone.utc).isoformat()
            }
            provisioned.append({
                "module_key": mod_key,
                "thread_id": new_thread_id,
                "title": topic.get("title")
            })
            modified = True

    if modified:
        with open(registry_file, "w", encoding="utf-8") as f:
            json.dump(registry, f, ensure_ascii=False, indent=2)
        return {
            "action": "RECONCILED",
            "reason": "topics_provisioned",
            "provisioned": provisioned
        }
    else:
        return {
            "action": "NOOP",
            "reason": "all_topics_registered"
        }


def main():
    parser = argparse.ArgumentParser(description="Reconcile Task Loop Topics")
    parser.add_argument("--root", default=".", help="Project root")
    parser.add_argument("--manifest", required=True, help="topics.json path")
    parser.add_argument("--registry", required=True, help="sessions.json path")
    args = parser.parse_args()

    result = reconcile_topics(args.root, args.manifest, args.registry)
    print(json.dumps(result, ensure_ascii=False))
    if result["action"] == "INVALID":
        sys.exit(2)


if __name__ == "__main__":
    main()

File: scripts/reconcile_task_loop_topics.js
#!/usr/bin/env node
/**
 * Reconcile Task Loop Topics -- [LEGACY v1 工具] 期望 topics[].module_key / registry.modules[key].thread_id 旧结构, 与 Schema v4 不兼容; 对现代状态文件天然 NOOP。保留仅作历史参考, 请勿在新流程使用 (改用 init/new-session)。 (Node.js)
 * Auto-provisions topic thread sessions for enabled topics declared in topics.json.
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

function reconcileTopics(projectRoot, manifestPath, registryPath) {
  const root = path.resolve(projectRoot);
  const manifestFile = path.isAbsolute(manifestPath) ? manifestPath : path.join(root, manifestPath);
  const registryFile = path.isAbsolute(registryPath) ? registryPath : path.join(root, registryPath);

  if (!fs.existsSync(manifestFile)) {
    return { action: 'INVALID', reason: `Manifest file not found: ${manifestFile}` };
  }
  if (!fs.existsSync(registryFile)) {
    return { action: 'INVALID', reason: `Registry file not found: ${registryFile}` };
  }

  let manifest, registry;
  try {
    manifest = JSON.parse(fs.readFileSync(manifestFile, 'utf8'));
    registry = JSON.parse(fs.readFileSync(registryFile, 'utf8'));
  } catch (e) {
    return { action: 'INVALID', reason: 'json_parse_error', message: e.message };
  }

  if (!registry.modules) {
    registry.modules = {};
  }

  const provisioned = [];
  let modified = false;

  for (const topic of (manifest.topics || [])) {
    if (!topic || topic.enabled === false) continue;

    const modKey = topic.module_key;
    if (!modKey) continue;

    const regMod = registry.modules[modKey];
    if (!regMod || !regMod.thread_id) {
      const newThreadId = crypto.randomUUID();
      registry.modules[modKey] = {
        thread_id: newThreadId,
        title: topic.title || `Topic ${modKey}`,
        title_prefix: topic.title_prefix || `${modKey}-`,
        memory_docs: topic.memory_docs || [],
        auto_provisioned: true,
        provisioned_at: new Date().toISOString()
      };
      provisioned.push({
        module_key: modKey,
        thread_id: newThreadId,
        title: topic.title
      });
      modified = true;
    }
  }

  if (modified) {
    fs.writeFileSync(registryFile, JSON.stringify(registry, null, 2), 'utf8');
    return {
      action: 'RECONCILED',
      reason: 'topics_provisioned',
      provisioned: provisioned
    };
  } else {
    return {
      action: 'NOOP',
      reason: 'all_topics_registered'
    };
  }
}

function main() {
  const args = process.argv.slice(2);
  let root = '.';
  let manifest = null;
  let registry = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--root' && args[i + 1]) root = args[++i];
    else if (args[i] === '--manifest' && args[i + 1]) manifest = args[++i];
    else if (args[i] === '--registry' && args[i + 1]) registry = args[++i];
  }

  if (!manifest || !registry) {
    console.error('Missing required arguments: --manifest, --registry');
    process.exit(1);
  }

  const result = reconcileTopics(root, manifest, registry);
  console.log(JSON.stringify(result));
  if (result.action === 'INVALID') process.exit(2);
}

if (require.main === module) {
  main();
}

module.exports = { reconcileTopics };
